apps/hes/fsw/src/hes_app.c

```diff
@@ -0,0 +1,1012 @@
+                    break;
+            }
+        }
+        else if (iStatus == CFE_SB_NO_MESSAGE)
+        {
+            break;
+        }
+        else
+        {
+            (void) CFE_EVS_SendEvent(HES_PIPE_ERR_EID, CFE_EVS_ERROR,
+                  "Data pipe read error (0x%08X)", (unsigned int)iStatus);
+            HES_AppData.uiRunStatus = CFE_ES_APP_ERROR;
+            break;
+        }
+    }
+}
+
+/* Conjugate with inversed quaternion */
+/* Body to NED Frame*/
+void Quaternion_Conjugate(const float* v, const float* data, float* result)
+{
+    float q0q0 = data[0] * data[0];
+    float q1q1 = data[1] * data[1];
+    float q2q2 = data[2] * data[2];
+    float q3q3 = data[3] * data[3];
+    
+    result[0] = (v[0] * (
+            q0q0 + q1q1 - q2q2 - q3q3) +
+            v[1] * 2.0f * (data[1] * data[2] - data[0] * data[3]) +
+            v[2] * 2.0f * (data[0] * data[2] + data[1] * data[3]));
+
+    result[1] = (v[0] * 2.0f * (data[1] * data[2] + data[0] * data[3]) +
+            v[1] * (q0q0 - q1q1 + q2q2 - q3q3) +
+            v[2] * 2.0f * (data[2] * data[3] - data[0] * data[1]));
+
+    result[2] = (v[0] * 2.0f * (data[1] * data[3] - data[0] * data[2]) +
+            v[1] * 2.0f * (data[0] * data[1] + data[2] * data[3]) +
+            v[2] * (q0q0 - q1q1 - q2q2 + q3q3));
+}
+
+
+/* Conjugate with inversed quaternion */
+/* NED to Body Frame*/
+void Quaternion_ConjugateInversed(const float* v, const float* data, float* result)
+{
+    float q0q0 = data[0] * data[0];
+    float q1q1 = data[1] * data[1];
+    float q2q2 = data[2] * data[2];
+    float q3q3 = data[3] * data[3];
+    
+    result[0] = (v[0] * (
+            q0q0 + q1q1 - q2q2 - q3q3) +
+            v[1] * 2.0f * (data[1] * data[2] + data[0] * data[3]) +
+            v[2] * 2.0f * (data[1] * data[3] - data[0] * data[2]));
+
+    result[1] = (v[0] * 2.0f * (data[1] * data[2] - data[0] * data[3]) +
+            v[1] * (q0q0 - q1q1 + q2q2 - q3q3) +
+            v[2] * 2.0f * (data[2] * data[3] + data[0] * data[1]));
+
+    result[2] = (v[0] * 2.0f * (data[1] * data[3] + data[0] * data[2]) +
+            v[1] * 2.0f * (data[2] * data[3] - data[0] * data[1]) +
+            v[2] * (q0q0 - q1q1 - q2q2 + q3q3));
+}
+
+
+void Quaternion_RotationMatrix(float* data, float** R)
+{
+	float aSq = data[0] * data[0];
+	float bSq = data[1] * data[1];
+	float cSq = data[2] * data[2];
+	float dSq = data[3] * data[3];
+	R[0][0] = aSq + bSq - cSq - dSq;
+	R[0][1] = 2.0f * (data[1] * data[2] - data[0] * data[3]);
+	R[0][2] = 2.0f * (data[0] * data[2] + data[1] * data[3]);
+	R[1][0] = 2.0f * (data[1] * data[2] + data[0] * data[3]);
+	R[1][1] = aSq - bSq + cSq - dSq;
+	R[1][2] = 2.0f * (data[2] * data[3] - data[0] * data[1]);
+	R[2][0] = 2.0f * (data[1] * data[3] - data[0] * data[2]);
+	R[2][1] = 2.0f * (data[0] * data[1] + data[2] * data[3]);
+	R[2][2] = aSq - bSq - cSq + dSq;
+}
+
+void Quaternion_ToEuler(float* data, float* euler)
```

Comment:
We already have this function (Vector3F Quaternion::ToEuler(void) const) defined in "apps/px4lib/fsw/src/math/Quaternion.cpp". It is fully unit tested. Use it instead.